Add support for Tuya Smart Air Housekeeper 6in1 Air Quality Monitor

m4k2k opened the pull request:
Goal: Add support for [Tuya Smart Air Housekeeper](https://zigbee.blakadder.com/Tuya_DCR-KQG.html) (aka TS0601 | _TZE200_dwcarsat)

This PR is a part of series of PR's affecting deconz-rest-plugin, pydeconz, homeassistant.
List: 

The commit [61a1fdd](https://github.com/Kane610/deconz/commit/61a1fdd51e3e9c29d9855caca8e1db758f4ed628) is required to prevent pydeconz from crashing if the airquality value is none - the sensor doesn't support it.
In HA this will be used in combination with commit [f9bc6d](https://github.com/Kane610/deconz/commit/f9bca6db80fb5e19871bbea1f210cfd70a556188) "supports airquality" check to hide the empty airquality sensor entity.

Result in HA once all PR's are merged:

m4k2k commented:
Test session starts (platform: win32, Python 3.10.10, pytest 7.2.1, pytest-sugar 0.9.5)
rootdir: D:\pydeconz, configfile: pyproject.toml
plugins: anyio-3.6.2, aiohttp-1.0.4, asyncio-0.20.2, cov-3.0.0, forked-1.4.0, pytest_freezer-0.4.6, picked-0.4.6, socket-0.5.1, sugar-0.9.5, test-groups-1.0.3, timeout-2.1.0, unordered-0.5.2, xdist-2.5.0, requests-mock-1.10.0, respx-0.20.1
asyncio: mode=auto
collecting ... 
 tests/test_alarm_system.py ✓                                                                                                                                                                                                                                                                      1% ▏         
 tests/test_api.py ✓✓                                                                                                                                                                                                                                                                              2% ▎         
 tests/test_config.py ✓                                                                                                                                                                                                                                                                            3% ▍
 tests/test_events.py ✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓                                                                                                                                                                                                                                                        16% █▋        
 tests/test_gateway.py ✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓                                                                                                                                                                                                                                                     30% ███       
 tests/test_groups.py ✓✓✓✓✓✓✓                                                                                                                                                                                                                                                                     35% ███▌      
 tests/test_lights.py ✓                                                                                                                                                                                                                                                                           36% ███▋      
 tests/test_scenes.py ✓✓                                                                                                                                                                                                                                                                          37% ███▊      
 tests/test_sensors.py ✓                                                                                                                                                                                                                                                                          38% ███▊      
 tests/test_utils.py ✓✓✓✓✓✓✓✓✓                                                                                                                                                                                                                                                                    43% ████▍     
 tests/lights/test_configuration_tool.py ✓                                                                                                                                                                                                                                                        44% ████▌     
 tests/lights/test_cover.py ✓✓✓✓                                                                                                                                                                                                                                                                  47% ████▋     
 tests/lights/test_light.py ✓✓✓✓✓✓✓✓✓                                                                                                                                                                                                                                                             53% █████▍    
 tests/lights/test_lock.py ✓✓                                                                                                                                                                                                                                                                     54% █████▌    
 tests/lights/test_range_extender.py ✓                                                                                                                                                                                                                                                            55% █████▌    
 tests/lights/test_siren.py ✓✓                                                                                                                                                                                                                                                                    56% █████▋    
 tests/sensors/test_air_purifier.py ✓✓                                                                                                                                                                                                                                                            57% █████▊    
 tests/sensors/test_air_quality.py ✓✓✓✓                                                                                                                                                                                                                                                           60% ██████    
 tests/sensors/test_alarm.py ✓                                                                                                                                                                                                                                                                    61% ██████▏   
 tests/sensors/test_ancillary_control.py ✓✓                                                                                                                                                                                                                                                       62% ██████▎   
 tests/sensors/test_battery.py ✓                                                                                                                                                                                                                                                                  62% ██████▍   
 tests/sensors/test_carbon_monoxide.py ✓                                                                                                                                                                                                                                                          63% ██████▍   
 tests/sensors/test_consumption.py ✓                                                                                                                                                                                                                                                              64% ██████▍   
 tests/sensors/test_daylight.py ✓✓                                                                                                                                                                                                                                                                65% ██████▌   
 tests/sensors/test_door_lock.py ✓✓                                                                                                                                                                                                                                                               66% ██████▋   
 tests/sensors/test_fire.py ✓✓                                                                                                                                                                                                                                                                    68% ██████▊   
 tests/sensors/test_generic_flag.py ✓                                                                                                                                                                                                                                                             68% ██████▉   
 tests/sensors/test_generic_status.py ✓                                                                                                                                                                                                                                                           69% ██████▉   
 tests/sensors/test_humidity.py ✓✓                                                                                                                                                                                                                                                                70% ███████▏  
 tests/sensors/test_light_level.py ✓✓                                                                                                                                                                                                                                                             72% ███████▎  
 tests/sensors/test_moisture.py ✓                                                                                                                                                                                                                                                                 72% ███████▎  
 tests/sensors/test_open_close.py ✓                                                                                                                                                                                                                                                               73% ███████▍  
 tests/sensors/test_power.py ✓✓                                                                                                                                                                                                                                                                   74% ███████▌  
 tests/sensors/test_presence.py ✓✓✓                                                                                                                                                                                                                                                               76% ███████▋  
 tests/sensors/test_pressure.py ✓                                                                                                                                                                                                                                                                 77% ███████▊  
 tests/sensors/test_relative_rotary.py ✓                                                                                                                                                                                                                                                          78% ███████▊  
 tests/sensors/test_switch.py ✓✓✓✓✓✓                                                                                                                                                                                                                                                              82% ████████▎ 
 tests/sensors/test_temperature.py ✓                                                                                                                                                                                                                                                              82% ████████▎ 
 tests/sensors/test_thermostat.py ✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓                                                                                                                                                                                                                                        98% █████████▊
 tests/sensors/test_time.py ✓                                                                                                                                                                                                                                                                     99% █████████▉
 tests/sensors/test_vibration.py ✓                                                                                                                                                                                                                                                                99% █████████▉
 tests/sensors/test_water.py ✓                                                                                                                                                                                                                                                                   100% ██████████

Results (1.42s):
     152 passed

m4k2k commented:
Hello @Kane610,
I had to remove the KeyError test as the error is not thrown anymore. https://github.com/Kane610/deconz/pull/408/commits/33a35764016ca47aafc33648d3359951e74af3fd
They 6in1 airquality box doesn't provide an 'airquality' value, just separate sensor data. Therefore I had to do https://github.com/Kane610/deconz/pull/408/commits/61a1fdd51e3e9c29d9855caca8e1db758f4ed628 to stop raising an error / crashing pydeconz and HA.

Kane610 commented:
Thanks! I'll try to have a look today or tomorrow

m4k2k commented:
PS D:\pydeconz> pre-commit run --all-files
black....................................................................Passed
isort....................................................................Passed

m4k2k commented:
Hello @Kane610
pre-commit depdency isort needs to bump from 5.11.4 to 5.11.5
https://github.com/Kane610/deconz/blob/0398874036e63995e7567ca70c42e85dbdb68296/.pre-commit-config.yaml#L10
because of a bug https://github.com/home-assistant/core/issues/86892
which is fixed in 5.11.5 https://github.com/home-assistant/core/issues/86892#issuecomment-1409991341
shall I add it to this PR or do you want a separate PR?

Kane610 commented:
> 

I think that is only an issue if you have shared python environments between pydeconz and home assistant

Kane610 commented on pydeconz/models/sensor/air_quality.py:
```diff
@@ -66,7 +68,7 @@ class AirQuality(SensorBase):
     def air_quality(self) -> str:  # AirQualityValue:
         """Air quality."""
-        return AirQualityValue(self.raw["state"]["airquality"]).value
+        return AirQualityValue(self.raw["state"].get("airquality")).value
```
This will cause the unknown logging from the air quality enum.
```suggestion
        return AirQualityValue(self.raw["state"].get("airquality", "unknown")).value
```
I'm not really happy with my suggested change but, lets keep it for now

m4k2k replied:
merged suggestions

Kane610 commented on pydeconz/models/sensor/air_quality.py:
```diff
@@ -78,6 +80,16 @@ def pm_2_5(self) -> int | None:
         """Air quality PM2.5 (µg/m³)."""
         return self.raw["state"].get("pm2_5")
 
+    @property
```
Sort properties by name

m4k2k replied:
done

Kane610 commented:
> Hello @Kane610,
> I had to remove the KeyError test as the error is not thrown anymore. [33a3576](https://github.com/Kane610/deconz/commit/33a35764016ca47aafc33648d3359951e74af3fd)
> The 6in1 airquality box doesn't provide an 'airquality' value, just separate sensor data. Therefore, I had to do [61a1fdd](https://github.com/Kane610/deconz/commit/61a1fdd51e3e9c29d9855caca8e1db758f4ed628) to stop raising an error / crashing pydeconz and HA.

Rather than changing the air_quality property what if you use the supports_air_quality instead of air_quality property on https://github.com/home-assistant/core/blob/8993a36e27dd40d99cb885e2eb415d5414ed4061/homeassistant/components/deconz/sensor.py#L111

Kane610 reviewed (changes requested):
Good job! Some smaller things to fix. Test coverage is 100% for air quality?

Kane610 commented on pydeconz/models/sensor/air_quality.py:
```diff
@@ -78,6 +80,16 @@ def pm_2_5(self) -> int | None:
         return self.raw["state"].get("pm2_5")
 
+    @property
+    def air_quality_formaldehyd(self) -> int | None:
```
Add supports_xxxxxx for the new properties as well

m4k2k replied:
done

darkstoneathome commented:
Hi @m4k2k , thanks for the good work! I am already very excited about this update. Does anyone know what is still missing here?

m4k2k commented on pydeconz/models/sensor/air_quality.py:
```diff
@@ -22,6 +22,8 @@ class TypedAirQualityState(TypedDict):
     ]
     airqualityppb: int
     pm2_5: int
+    airqualityformaldehyd: int
```
done

m4k2k commented:
Hello @Kane610, I fixed your suggestions.
The PR is now 19 commits behind the master.
Shall I merge the 19 commits first into my PR/master to check conflicts?
Is anything else required / blocking the merge?
best regards

m4k2k commented:
Hello @Kane610, I fixed your suggestions.
The PR is now 19 commits behind , shall I merge the 19 commits first into my master to check conflicts?

> > Hello @Kane610,
> > I had to remove the KeyError test as the error is not thrown anymore. [33a3576](https://github.com/Kane610/deconz/commit/33a35764016ca47aafc33648d3359951e74af3fd)
> > The 6in1 airquality box doesn't provide an 'airquality' value, just separate sensor data. Therefore, I had to do [61a1fdd](https://github.com/Kane610/deconz/commit/61a1fdd51e3e9c29d9855caca8e1db758f4ed628) to stop raising an error / crashing pydeconz and HA.
> 
> Rather than changing the air_quality property what if you use the supports_air_quality instead of air_quality property on https://github.com/home-assistant/core/blob/8993a36e27dd40d99cb885e2eb415d5414ed4061/homeassistant/components/deconz/sensor.py#L111

I had this implemented at first and it was working. Later I switched to the lamda solution as it's been used for the other properties as well. I thought the supports functions were deprecated as they were not used anymore.

Kane610 commented:
> Hello @Kane610, I fixed your suggestions.
> The PR is now 19 commits behind the master.
> Shall I merge the 19 commits first into my PR/master to check conflicts?
> Is anything else required / blocking the merge?
> best regards

Its always good to keep up with master, I do not think any changes are related as its either light enums updated or dependencies

m4k2k commented:
> > Hello @Kane610, I fixed your suggestions.
> > The PR is now 19 commits behind the master.
> > Shall I merge the 19 commits first into my PR/master to check conflicts?
> > Is anything else required / blocking the merge?
> > best regards
> 
> Its always good to keep up with master, I do not think any changes are related as its either light enums updated or dependencies

OK I do the merge later, I will also remove the blank / blocking flake8

m4k2k commented:
just verified with the latest changes, everything is still working

m4k2k commented:
test coverage back at 100% for air_quality.py

========================================================================== test session starts ==========================================================================
platform linux -- Python 3.10.10, pytest-7.2.2, pluggy-1.0.0
rootdir: /testing/pydeconz, configfile: pyproject.toml
plugins: aiohttp-1.0.4, asyncio-0.20.3, rerunfailures-11.1.2, mock-3.10.0, cov-4.0.0, anyio-3.6.2
asyncio: mode=auto
collected 4 items

tests/sensors/test_air_quality.py ....                                                                                                                            [100%]

---------- coverage: platform linux, python 3.10.10-final-0 ----------
Name                                    Stmts   Miss  Cover   Missing
---------------------------------------------------------------------
pydeconz/models/sensor/air_quality.py      57      0   100%
---------------------------------------------------------------------
TOTAL                                      57      0   100%


=========================================================================== 4 passed in 0.41s ===========================================================================

Kane610 commented on pydeconz/models/sensor/air_quality.py:
```diff
@@ -78,6 +90,21 @@ def pm_2_5(self) -> int | None:
+        return "airqualityco2" in self.raw["state"]
+
+    @property
+    def supports_air_quality_formaldehyd(self) -> bool:
```
```suggestion
    def supports_air_quality_formaldehyde(self) -> bool:
```
At the very least it can be properly spelled on pydeconz side :)

Kane610 commented on pydeconz/models/sensor/air_quality.py:
```diff
@@ -20,6 +20,8 @@ class TypedAirQualityState(TypedDict):
         "out of scale",
     ]
+    airqualityco2: int
+    airqualityformaldehyd: int
```
This is misspelled, I understand its nothing you can do anything about, I've reported it here https://github.com/dresden-elektronik/deconz-rest-plugin/issues/6810

m4k2k replied:
I'm afraid this happened by Denglisch - in german its formaldehyd. I check where this needs to be fixed in the repo.

Kane610 replied:
As I wrote, I've already reported it. Proposal: you add a second line with the correctly spelled airqualityformaldehyde: int as well, also in the property you try with both formats prioritising the proper spelling, then both will work

Kane610 replied:
This is the last thing then we can merge and make a release

m4k2k replied:
> As I wrote, I've already reported it. Proposal: you add a second line with the correctly spelled airqualityformaldehyde: int as well, also in the property you try with both formats prioritising the proper spelling, then both will work

while you were writing I directly adapted. I also changed directly to the new format introduced here [#6673#issuecomment-1475150705](https://github.com/dresden-elektronik/deconz-rest-plugin/pull/6673#issuecomment-1475150705)

```
state/airquality_co2_density
state/airquality_formaldehyde_density
```
hope this is ok - else I can undo.
As the HA PR [#87866](https://github.com/home-assistant/core/pull/87866) is not yet merged and the deconz rest api PR [#6673](https://github.com/dresden-elektronik/deconz-rest-plugin/pull/6673) is also pending, I think we can change as nothing is breaking. The DDF is not yet usable (except you build the rest plugin manually)

Kane610 replied:
Great! Kudos to you for being on top of all the changes! <3

I would like for us to wait a couple of days for https://github.com/dresden-elektronik/deconz-rest-plugin/pull/6673 to be merged before merging this one as that is still in flux. Lets aim to have it all merged prior to hass 2023.4 branches off in 10 days.

Kane610 reviewed (changes requested):
Thanks! Nearly there! Minor changes left.

Im thinking whether or not the supported properties are really relevant or not looking at it not being needed for ppl on hass side. I'll change that myself down the line if I change my mind.

m4k2k commented:
```
========================================================================= test session starts =========================================================================
platform linux -- Python 3.10.10, pytest-7.2.2, pluggy-1.0.0
rootdir: /testing/pydeconz, configfile: pyproject.toml
plugins: aiohttp-1.0.4, asyncio-0.21.0, rerunfailures-11.1.2, mock-3.10.0, cov-4.0.0, anyio-3.6.2
asyncio: mode=auto
collected 4 items

tests/sensors/test_air_quality.py ....                                                                                                                          [100%]

---------- coverage: platform linux, python 3.10.10-final-0 ----------
Name                                    Stmts   Miss  Cover   Missing
---------------------------------------------------------------------
pydeconz/models/sensor/air_quality.py      57      0   100%
---------------------------------------------------------------------
TOTAL                                      57      0   100%


========================================================================== 4 passed in 0.42s ==========================================================================
```

m4k2k commented:
```
bash-5.1# pre-commit run --all-files
ruff.....................................................................Passed
black....................................................................Passed
isort....................................................................Passed
mypy.....................................................................Passed
```

```
============================================================================================================================================= test session starts ==============================================================================================================================================
platform linux -- Python 3.10.10, pytest-7.2.2, pluggy-1.0.0
rootdir: /testing/pydeconz, configfile: pyproject.toml
plugins: aiohttp-1.0.4, asyncio-0.21.0, rerunfailures-11.1.2, mock-3.10.0, cov-4.0.0, anyio-3.6.2
asyncio: mode=auto
collected 152 items

tests/test_alarm_system.py .                                                                                                                                                                                                                                                                             [  0%]
tests/test_api.py ..                                                                                                                                                                                                                                                                                     [  1%]
tests/test_config.py .                                                                                                                                                                                                                                                                                   [  2%]
tests/test_events.py ....................                                                                                                                                                                                                                                                                [ 15%]
tests/test_gateway.py ......................                                                                                                                                                                                                                                                             [ 30%]
tests/test_groups.py .......                                                                                                                                                                                                                                                                             [ 34%]
tests/test_lights.py .                                                                                                                                                                                                                                                                                   [ 35%]
tests/test_scenes.py ..                                                                                                                                                                                                                                                                                  [ 36%]
tests/test_sensors.py .                                                                                                                                                                                                                                                                                  [ 37%]
tests/test_utils.py .........                                                                                                                                                                                                                                                                            [ 43%]
tests/lights/test_configuration_tool.py .                                                                                                                                                                                                                                                                [ 44%]
tests/lights/test_cover.py ....                                                                                                                                                                                                                                                                          [ 46%]
tests/lights/test_light.py .........                                                                                                                                                                                                                                                                     [ 52%]
tests/lights/test_lock.py ..                                                                                                                                                                                                                                                                             [ 53%]
tests/lights/test_range_extender.py .                                                                                                                                                                                                                                                                    [ 54%]
tests/lights/test_siren.py ..                                                                                                                                                                                                                                                                            [ 55%]
tests/sensors/test_air_purifier.py ..                                                                                                                                                                                                                                                                    [ 57%]
tests/sensors/test_air_quality.py ....                                                                                                                                                                                                                                                                   [ 59%]
tests/sensors/test_alarm.py .                                                                                                                                                                                                                                                                            [ 60%]
tests/sensors/test_ancillary_control.py ..                                                                                                                                                                                                                                                               [ 61%]
tests/sensors/test_battery.py .                                                                                                                                                                                                                                                                          [ 62%]
tests/sensors/test_carbon_monoxide.py .                                                                                                                                                                                                                                                                  [ 63%]
tests/sensors/test_consumption.py .                                                                                                                                                                                                                                                                      [ 63%]
tests/sensors/test_daylight.py ..                                                                                                                                                                                                                                                                        [ 65%]
tests/sensors/test_door_lock.py ..                                                                                                                                                                                                                                                                       [ 66%]
tests/sensors/test_fire.py ..                                                                                                                                                                                                                                                                            [ 67%]
tests/sensors/test_generic_flag.py .                                                                                                                                                                                                                                                                     [ 68%]
tests/sensors/test_generic_status.py .                                                                                                                                                                                                                                                                   [ 69%]
tests/sensors/test_humidity.py ..                                                                                                                                                                                                                                                                        [ 70%]
tests/sensors/test_light_level.py ..                                                                                                                                                                                                                                                                     [ 71%]
tests/sensors/test_moisture.py .                                                                                                                                                                                                                                                                         [ 72%]
tests/sensors/test_open_close.py .                                                                                                                                                                                                                                                                       [ 73%]
tests/sensors/test_power.py ..                                                                                                                                                                                                                                                                           [ 74%]
tests/sensors/test_presence.py ...                                                                                                                                                                                                                                                                       [ 76%]
tests/sensors/test_pressure.py .                                                                                                                                                                                                                                                                         [ 76%]
tests/sensors/test_relative_rotary.py .                                                                                                                                                                                                                                                                  [ 77%]
tests/sensors/test_switch.py ......                                                                                                                                                                                                                                                                      [ 81%]
tests/sensors/test_temperature.py .                                                                                                                                                                                                                                                                      [ 82%]
tests/sensors/test_thermostat.py ........................                                                                                                                                                                                                                                                [ 98%]
tests/sensors/test_time.py .                                                                                                                                                                                                                                                                             [ 98%]
tests/sensors/test_vibration.py .                                                                                                                                                                                                                                                                        [ 99%]
tests/sensors/test_water.py .                                                                                                                                                                                                                                                                            [100%]

---------- coverage: platform linux, python 3.10.10-final-0 ----------
Name                                          Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------
pydeconz/__init__.py                              2      0   100%
pydeconz/__main__.py                             57     57     0%   3-107
pydeconz/config.py                              131      0   100%
pydeconz/errors.py                               17      0   100%
pydeconz/gateway.py                             104      0   100%
pydeconz/interfaces/__init__.py                   0      0   100%
pydeconz/interfaces/alarm_systems.py             26      0   100%
pydeconz/interfaces/api_handlers.py             125      0   100%
pydeconz/interfaces/events.py                    29      0   100%
pydeconz/interfaces/groups.py                    19      0   100%
pydeconz/interfaces/lights.py                    90      0   100%
pydeconz/interfaces/scenes.py                    25      0   100%
pydeconz/interfaces/sensors.py                  209      0   100%
pydeconz/models/__init__.py                      74      0   100%
pydeconz/models/alarm_system.py                 115      0   100%
pydeconz/models/api.py                           38      0   100%
pydeconz/models/deconz_device.py                 30      0   100%
pydeconz/models/event.py                         51      0   100%
pydeconz/models/group.py                        111      0   100%
pydeconz/models/light/__init__.py                12      0   100%
pydeconz/models/light/configuration_tool.py       2      0   100%
pydeconz/models/light/cover.py                   37      0   100%
pydeconz/models/light/light.py                  159      0   100%
pydeconz/models/light/lock.py                    11      0   100%
pydeconz/models/light/range_extender.py           2      0   100%
pydeconz/models/light/siren.py                   12      0   100%
pydeconz/models/scene.py                         38      0   100%
pydeconz/models/sensor/__init__.py               34      0   100%
pydeconz/models/sensor/air_purifier.py           52      0   100%
pydeconz/models/sensor/air_quality.py            45      0   100%
pydeconz/models/sensor/alarm.py                  11      0   100%
pydeconz/models/sensor/ancillary_control.py      51      0   100%
pydeconz/models/sensor/battery.py                11      0   100%
pydeconz/models/sensor/carbon_monoxide.py        11      0   100%
pydeconz/models/sensor/consumption.py            18      0   100%
pydeconz/models/sensor/daylight.py               68      0   100%
pydeconz/models/sensor/door_lock.py              26      0   100%
pydeconz/models/sensor/fire.py                   15      0   100%
pydeconz/models/sensor/generic_flag.py           11      0   100%
pydeconz/models/sensor/generic_status.py         11      0   100%
pydeconz/models/sensor/humidity.py               20      0   100%
pydeconz/models/sensor/light_level.py            37      0   100%
pydeconz/models/sensor/moisture.py               11      0   100%
pydeconz/models/sensor/open_close.py             11      0   100%
pydeconz/models/sensor/power.py                  19      0   100%
pydeconz/models/sensor/presence.py               77      0   100%
pydeconz/models/sensor/pressure.py               11      0   100%
pydeconz/models/sensor/relative_rotary.py        23      0   100%
pydeconz/models/sensor/switch.py                 68      0   100%
pydeconz/models/sensor/temperature.py            14      0   100%
pydeconz/models/sensor/thermostat.py            196      0   100%
pydeconz/models/sensor/time.py                   11      0   100%
pydeconz/models/sensor/vibration.py              33      0   100%
pydeconz/models/sensor/water.py                  11      0   100%
pydeconz/utils.py                                58      0   100%
pydeconz/websocket.py                            93     62    33%   46-55, 65-67, 72-75, 80, 84-85, 89, 95, 99-140, 144-145, 153-168
---------------------------------------------------------------------------
TOTAL                                          2583    119    95%


============================================================================================================================================= 152 passed in 13.25s =============================================================================================================================================

```

updated the HA and deconz branch to "latest"

sensors are all still up and running :)
![image](https://user-images.githubusercontent.com/7247710/227729229-507b46ba-20b8-433e-9af4-2d0b684c2440.png)

all tests, precommit, system test done

anything open todo? @Kane610

Kane610 commented:
Looked over it and it looks good, gonna have a once over tomorrow again and then we can merge, I expect we can get everything in before beta cut off on Wednesday.

Kane610 reviewed (approved):
I have some thoughts still, but it's perfectly fine to release. Thanks!

Kane610 commented:
I will build a release tonight probably. I'll merge it in time for that, if you have any final things left just ping here.

m4k2k commented:
> I will build a release tonight probably. I'll merge it in time for that, if you have any final things left just ping here.

No leftovers, ready to merge :)
